roles/ansible-test-splitter/files/split_targets.py

```diff
@@ -6,52 +6,116 @@
-for target in targets.glob("*"):
-    aliases = target / "aliases"
-    if not target.is_dir():
-        continue
-    if not aliases.is_file():
-        continue
-    lines = aliases.read_text().split("\n")
-    combined = lines + [target.name]
-    if targets_from_cli and not set(combined) & set(targets_from_cli):
-        continue
+
+
+def get_job_list(prefix, total_jobs):
+    return [f"{prefix}{i}" for i in range(total_jobs)]
+
+
+def to_skip_because_disabled(lines):
     if "disabled" in lines:
-        continue
+        return True
     if "unstable" in lines:
-        continue
-    if "slow" in lines or "# reason: slow" in lines:
-        batches.append([target.name])
-    else:
-        regular_targets.append(target.name)
-
-remaining_jobs = max(total_jobs - len(batches), 1)
-
-for x in range(remaining_jobs):
-    batch = regular_targets[x::remaining_jobs]
-    if batch:
-        batches.append(batch)
-
-result = {
-    "data": {
-        "zuul": {"child_jobs": jobs[0:len(batches)]},
-        "child": {"targets_to_test": batches},
+        return True
+    return False
+
+
+def is_slow(lines):
+    if "slow" in lines:
+        return True
+    if "# reason: slow" in lines:
+        return True
+    return False
+
+
+def iter_target_dir(collection_path):
+    targets = PosixPath(os.path.join(collection_path, "tests/integration/targets/"))
+    for target in targets.glob("*"):
+        aliases = target / "aliases"
+        if not target.is_dir():
+            continue
+        if not aliases.is_file():
+            continue
+        lines = aliases.read_text().split("\n")
+        yield (target, lines)
+
+
+def to_skip_because_of_targets_parameters(target_name, lines, targets_from_cli):
+    """--targets parameter is in use, we skip the targets not part of the list."""
+    combined = lines + [target_name]
+    if targets_from_cli and not set(combined) & set(targets_from_cli):
+        return True
+    return False
+
+
+def get_targets_to_run(collection_path, targets_from_cli):
+    slow_targets = []
+    regular_targets = []
+    for target, lines in iter_target_dir(collection_path):
+        if to_skip_because_of_targets_parameters(target.name, lines, targets_from_cli):
+            continue
+        if to_skip_because_disabled(lines):
+            continue
+        if is_slow(lines):
+            slow_targets.append([target.name])
+        else:
+            regular_targets.append(target.name)
+    return slow_targets, regular_targets
+
+
+def build_up_batches(slow_targets, regular_targets, total_jobs):
+    remaining_jobs = max(total_jobs - len(slow_targets), 1)
+    batches = [[i] for i in slow_targets]
+    for x in range(remaining_jobs):
+        batch = regular_targets[x::remaining_jobs]
+        if batch:
+            batches.append(batch)
+    return batches
+
+
+def build_result_struct(jobs, batches):
+    result = {
+        "data": {
+            "zuul": {"child_jobs": jobs[0 : len(batches)]},
```

Comment:
What happens if `len(batches)` is longer than `len(jobs)`? Would that mean some targets don't get run?

Reply:
Yes, probably. The goal of this PR is to add test coverage. I suggest we merge this PR and start improving stuff. It will be easier to understand what we do.